syntax = "proto3";

import "proto/server.proto";

// The service for Paxos protocol communication among servers.
service Paxos {
    rpc Prepare (PrepareArgs) returns (PrepareReply) {}
    rpc Accept (AcceptArgs) returns (AcceptReply) {}
    rpc Decide (DecideArgs) returns (DecideReply) {}
}

// A PrepareArgs is a message server sent to server to invoke prepare stage.
message PrepareArgs {
    int32 Seq = 1;
    int32 N = 2;
    int32 Sender = 3;
    int32 Done = 4;
}

// A PrepareReply is a message server sent to server after it completes prepare stage.
message PrepareReply {
    bool OK = 1;
    int32 Na = 2;
    Op Va = 3;
    int32 Done = 4;
}

// An AcceptArgs is a message server sent to server to invoke accept stage.
message AcceptArgs {
    int32 Seq = 1;
    int32 N = 2;
    Op V = 3;
}

// An AcceptReply is a message server sent to server after it completes accept stage.
message AcceptReply {
    bool OK = 1;
    int32 N = 2;
}

// A DecideArgs is a message server sent to sever to invoke decide stage.
message DecideArgs {
    int32 Seq = 1;
    Op V = 2;
}

// A Decide Reply is a message server sent to server after it complete decide stage.
message DecideReply {
    bool OK = 1;
}

